// @openpesa/pesa-js/lib/modules/Pesa.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Pesa = void 0;
var axios_1 = require("axios");
var constants = require("constants");
var crypto = require("crypto");
/**
 * Pesa
 * @package openpesaa/pesa-js
 * @class Pesa
 */
var Pesa = /** @class */ (function () {
    /**
     * @constructor
     * @param {IPesaOptions} options
     * @param environment
     *
     */
    function Pesa(options, environment) {
        var _a, _b;
        this.BASE_DOMAIN = 'https://openapi.m-pesa.com/';
        this.AUTH_URL = 'ipg/v2/vodacomTZN/getSession/';
        this.TRANSACTION_ROUTES = {
            b2c: 'ipg/v2/vodacomTZN/b2cPayment/singleStage/',
            c2b: 'ipg/v2/vodacomTZN/c2bPayment/singleStage/',
            ddc: 'ipg/v2/vodacomTZN/directDebitCreation/',
            ddp: 'ipg/v2/vodacomTZN/directDebitPayment/',
            query: 'ipg/v2/vodacomTZN/queryTransactionStatus/',
            rt: 'ipg/v2/vodacomTZN/reversal/',
        };
        if (environment === 'production') {
            this.baseURL = this.BASE_DOMAIN + 'openapi/';
        }
        else {
            this.baseURL = this.BASE_DOMAIN + 'sandbox/';
        }
        this.options = options;
        if (['node', 'io.js'].includes((_a = process === null || process === void 0 ? void 0 : process.release) === null || _a === void 0 ? void 0 : _a.name) == false) {
            throw new Error("Unexpected process release name " + ((_b = process === null || process === void 0 ? void 0 : process.release) === null || _b === void 0 ? void 0 : _b.name) + ". This may indicate that the Mollie API client is integrated into a website or app. This is not recommended, please see https://openpesa.github.io/pesa-js/guide/usage-guide#. If this is a mistake, please let us know: https://github.com/openpesa/pesa-js/issues");
        }
        if (!options.api_key && !options.public_key) {
            throw new TypeError('Missing parameter "apiKey" or "publickey".');
        }
    }
    /**
     * getSessionId
     */
    Pesa.prototype.getSessionId = function () {
        var _a;
        return (_a = this.options.sessionId) !== null && _a !== void 0 ? _a : null;
    };
    /**
     * setSessionId
     */
    Pesa.prototype.setSessionId = function (sessId) {
        this.options.sessionId = sessId;
    };
    /**
     * Generate a Session Key
     *
     * Before you can integrate on the M-Pesa OpenAPI solution, you must exchange your Application Key for a Session Key. The API Key is created with the creation of a new application. The Session Key acts as an access token that authorises the rest of your REST API calls to the system. A valid Session Key is needed to transact on M-Pesa using OpenAPI.
     *
     * Logic
     * Will return saved sessionId otherise get a new sessionId
     * Fetched sessionId will not be saved automatically
     *
     * @see {@link https://openapiportal.m-pesa.com/api-documentation#GeneratingtheSessionKey}
     * @returns {Promise}
     */
    Pesa.prototype.get_session = function () {
        return __awaiter(this, void 0, void 0, function () {
            var response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!!this.options.sessionId) return [3 /*break*/, 2];
                        return [4 /*yield*/, axios_1.default({
                                method: 'get',
                                url: this.baseURL + this.AUTH_URL,
                                headers: {
                                    Accept: 'application/json',
                                    Origin: '*',
                                    Authorization: 'Bearer ' + this.encrypt_key(this.options.api_key),
                                },
                            })];
                    case 1:
                        response = _a.sent();
                        return [2 /*return*/, __assign(__assign({}, response.data), { status: 200 })];
                    case 2: return [2 /*return*/, {
                            output_ResponseCode: '',
                            output_ResponseDesc: '',
                            output_TransactionID: '',
                            output_ConversationID: '',
                            output_ThirdPartyConversationID: '',
                            output_SessionID: this.options.sessionId,
                        }];
                }
            });
        });
    };
    /**
     *
     * Encrypting the API Key
     *
     * To generate your Session Key for the sandbox and live environments:
     * 1.    Log into OpenAPI with a developer account
     * 2.    On the APPLICATION page, click Create New Application. Creating an application will generate your unique API Application Key needed to authorise and authenticate your application on the server
     * 3.    Type your name and version number for the application and select the products you wish to use. (The application can be configured any time). Save your new application.
     * 4.    Copy and save the API Key.
     * 5.    Copy the Public Key from the below section.
     * 6.    Generate a decoded Base64 string from the Public Key
     * 7.    Generate an instance of an RSA cipher and use the Base 64 string as the input
     * 8.    Encode the API Key with the RSA cipher and digest as Base64 string format
     * 9.    The result is your encrypted API Key.
     *
     *
     * @see {@link https://openapiportal.m-pesa.com/api-documentation#GeneratingtheSessionKey }
     * @see {@link https://nodejs.org/api/crypto.html#crypto_crypto_createpublickey_key }
     *
     * @param {string} key
     * @return {string}
     */
    Pesa.prototype.encrypt_key = function (key) {
        var pk = '-----BEGIN PUBLIC KEY-----\n' + this.options.public_key + '\n' + '-----END PUBLIC KEY-----';
        return crypto
            .publicEncrypt({ key: pk, padding: constants.RSA_PKCS1_PADDING }, Buffer.from(key))
            .toString('base64');
    };
    /**
     * The Query Transaction Status
     * @name query
     * @function
     * @description The Query Transaction Status API call is used to query the status of the transaction that has been initiated.
     * @param {query} data
     * @returns {Promise}
     */
    Pesa.prototype.query = function (data, sessionID) {
        return __awaiter(this, void 0, void 0, function () {
            var output_SessionID, res, response, error_1, serverError;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 4, , 5]);
                        output_SessionID = '';
                        if (!!sessionID) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.get_session()];
                    case 1:
                        res = _a.sent();
                        output_SessionID = res.output_SessionID;
                        _a.label = 2;
                    case 2: return [4 /*yield*/, axios_1.default({
                            method: 'get',
                            url: this.baseURL + this.TRANSACTION_ROUTES.query,
                            headers: {
                                Accept: 'application/json',
                                Origin: '*',
                                Authorization: 'Bearer ' + this.encrypt_key(output_SessionID),
                            },
                            data: {
                                input_QueryReference: data.input_QueryReference,
                                input_ServiceProviderCode: data.input_ServiceProviderCode,
                                input_ThirdPartyConversationID: data.input_ThirdPartyConversationID,
                                input_Country: data.input_Country,
                            },
                        })];
                    case 3:
                        response = _a.sent();
                        return [2 /*return*/, __assign(__assign({}, response.data), { status: 200 })];
                    case 4:
                        error_1 = _a.sent();
                        serverError = error_1;
                        if (serverError === null || serverError === void 0 ? void 0 : serverError.response) {
                            return [2 /*return*/, __assign(__assign({}, serverError.response.data), { code: serverError.code, status: serverError.response.status })];
                        }
                        throw new Error((serverError === null || serverError === void 0 ? void 0 : serverError.message) || 'Internal server error');
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Customer to business (C2B)
     * @function c2b
     * @description The C2B API call is used as a standard customer-to-business transaction. Funds from the customer’s mobile money wallet will be deducted and be transferred to the mobile money wallet of the business. To authenticate and authorize this transaction, M-Pesa Payments Gateway will initiate a USSD Push message to the customer to gather and verify the mobile money PIN number. This number is not stored and is used only to authorize the transaction.
     * @api
     * @param {c2b} data
     *
     * @returns {Promise} Promise

     */
    Pesa.prototype.c2b = function (data, sessionID) {
        return __awaiter(this, void 0, void 0, function () {
            var output_SessionID, res, response, error_2, serverError;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 4, , 5]);
                        output_SessionID = '';
                        if (!!sessionID) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.get_session()];
                    case 1:
                        res = _a.sent();
                        output_SessionID = res.output_SessionID;
                        _a.label = 2;
                    case 2: return [4 /*yield*/, axios_1.default({
                            method: 'post',
                            url: this.baseURL + this.TRANSACTION_ROUTES.c2b,
                            headers: {
                                Accept: 'application/json',
                                Origin: '*',
                                Authorization: 'Bearer ' + this.encrypt_key(output_SessionID),
                            },
                            data: {
                                input_Amount: data.input_Amount,
                                input_Country: data.input_Country,
                                input_Currency: data.input_Currency,
                                input_CustomerMSISDN: data.input_CustomerMSISDN,
                                input_ServiceProviderCode: data.input_ServiceProviderCode,
                                input_ThirdPartyConversationID: data.input_ThirdPartyConversationID,
                                input_TransactionReference: data.input_TransactionReference,
                                input_PurchasedItemsDesc: data.input_PurchasedItemsDesc,
                            },
                        })];
                    case 3:
                        response = _a.sent();
                        return [2 /*return*/, __assign(__assign({}, response.data), { status: 200 })];
                    case 4:
                        error_2 = _a.sent();
                        serverError = error_2;
                        if (serverError === null || serverError === void 0 ? void 0 : serverError.response) {
                            return [2 /*return*/, __assign(__assign({}, serverError.response.data), { code: serverError.code, status: serverError.response.status })];
                        }
                        throw new Error((serverError === null || serverError === void 0 ? void 0 : serverError.message) || 'Internal server error');
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Business to Customer (B2C)
     *
     * The B2C API Call is used as a standard business-to-customer funds disbursement. Funds from the business account’s wallet will be deducted and paid to the mobile money wallet of the customer. Use cases for the B2C includes:
     *  -    Salary payments
     *  -    Funds transfers from business
     *  -    Charity pay-out
     *
     * @api
     * @param {b2c} data

     * @returns {Promise} Promise
     */
    Pesa.prototype.b2c = function (data, sessionID) {
        return __awaiter(this, void 0, void 0, function () {
            var output_SessionID, res, response, error_3, serverError;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 4, , 5]);
                        output_SessionID = '';
                        if (!!sessionID) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.get_session()];
                    case 1:
                        res = _a.sent();
                        output_SessionID = res.output_SessionID;
                        _a.label = 2;
                    case 2: return [4 /*yield*/, axios_1.default({
                            method: 'post',
                            url: this.baseURL + this.TRANSACTION_ROUTES.c2b,
                            headers: {
                                Accept: 'application/json',
                                Origin: '*',
                                Authorization: 'Bearer ' + this.encrypt_key(output_SessionID),
                            },
                            data: {
                                input_Amount: data.input_Amount,
                                input_Country: data.input_Country,
                                input_Currency: data.input_Currency,
                                input_CustomerMSISDN: data.input_CustomerMSISDN,
                                input_ServiceProviderCode: data.input_ServiceProviderCode,
                                input_ThirdPartyConversationID: data.input_ThirdPartyConversationID,
                                input_TransactionReference: data.input_TransactionReference,
                                input_PurchasedItemsDesc: data.input_PurchasedItemsDesc,
                            },
                        })];
                    case 3:
                        response = _a.sent();
                        return [2 /*return*/, __assign(__assign({}, response.data), { status: 200 })];
                    case 4:
                        error_3 = _a.sent();
                        serverError = error_3;
                        if (serverError === null || serverError === void 0 ? void 0 : serverError.response) {
                            return [2 /*return*/, __assign(__assign({}, serverError.response.data), { code: serverError.code, status: serverError.response.status })];
                        }
                        throw new Error((serverError === null || serverError === void 0 ? void 0 : serverError.message) || 'Internal server error');
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Business to business (B2B)
     *
     * The B2B API Call is used for business-to-business transactions. Funds from the business’ mobile money wallet will be deducted and transferred to the mobile money wallet of the other business. Use cases for the B2C includes:
     *  -  Stock purchases
     *  -  Bill payment
     *  -  Ad-hoc payment
     *
     * @api
     * @param data
     *
     * @returns {Promise} Promise
     */
    Pesa.prototype.b2b = function (data, sessionID) {
        return __awaiter(this, void 0, void 0, function () {
            var output_SessionID, res, response, error_4, serverError;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 4, , 5]);
                        output_SessionID = '';
                        if (!!sessionID) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.get_session()];
                    case 1:
                        res = _a.sent();
                        output_SessionID = res.output_SessionID;
                        _a.label = 2;
                    case 2: return [4 /*yield*/, axios_1.default({
                            method: 'post',
                            url: this.baseURL + this.TRANSACTION_ROUTES.c2b,
                            headers: {
                                Accept: 'application/json',
                                Origin: '*',
                                Authorization: 'Bearer ' + this.encrypt_key(output_SessionID),
                            },
                            data: {
                                input_Amount: data.input_Amount,
                                input_Country: data.input_Country,
                                input_Currency: data.input_Currency,
                                input_PrimaryPartyCode: data.input_PrimaryPartyCode,
                                input_ReceiverPartyCode: data.input_ReceiverPartyCode,
                                input_ThirdPartyConversationID: data.input_ThirdPartyConversationID,
                                input_TransactionReference: data.input_TransactionReference,
                                input_PurchasedItemsDesc: data.input_PurchasedItemsDesc,
                            },
                        })];
                    case 3:
                        response = _a.sent();
                        return [2 /*return*/, __assign(__assign({}, response.data), { status: 200 })];
                    case 4:
                        error_4 = _a.sent();
                        serverError = error_4;
                        if (serverError === null || serverError === void 0 ? void 0 : serverError.response) {
                            return [2 /*return*/, __assign(__assign({}, serverError.response.data), { code: serverError.code, status: serverError.response.status })];
                        }
                        throw new Error((serverError === null || serverError === void 0 ? void 0 : serverError.message) || 'Internal server error');
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    /**
     * Payment reversals
     *
     * The Reversal API is used to reverse a successful transaction. Using the Transaction ID of a previously successful transaction,  the OpenAPI will withdraw the funds from the recipient party’s mobile money wallet and revert the funds to the mobile money wallet of the initiating party of the original transaction.
     *
     * @api
     * @param data
     *
     * @returns {Promise} Promise
     *
     */
    Pesa.prototype.reverse = function (data, sessionID) {
        return __awaiter(this, void 0, void 0, function () {
            var output_SessionID, res, response, error_5, serverError;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 4, , 5]);
                        output_SessionID = '';
                        if (!!sessionID) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.get_session()];
                    case 1:
                        res = _a.sent();
                        output_SessionID = res.output_SessionID;
                        _a.label = 2;
                    case 2: return [4 /*yield*/, axios_1.default({
                            method: 'post',
                            url: this.baseURL + this.TRANSACTION_ROUTES.c2b,
                            headers: {
                                Accept: 'application/json',
                                Origin: '*',
                                Authorization: 'Bearer ' + this.encrypt_key(output_SessionID),
                            },
                            data: {
                                input_ReversalAmount: data.input_ReversalAmount,
                                input_TransactionID: data.input_TransactionID,
                                input_Country: data.input_Country,
                                input_ServiceProviderCode: data.input_ServiceProviderCode,
                                input_ThirdPartyConversationID: data.input_ThirdPartyConversationID,
                            },
                        })];
                    case 3:
                        response = _a.sent();
                        return [2 /*return*/, __assign(__assign({}, response.data), { status: 200 })];
                    case 4:
                        error_5 = _a.sent();
                        serverError = error_5;
                        if (serverError === null || serverError === void 0 ? void 0 : serverError.response) {
                            return [2 /*return*/, __assign(__assign({}, serverError.response.data), { code: serverError.code, status: serverError.response.status })];
                        }
                        throw new Error((serverError === null || serverError === void 0 ? void 0 : serverError.message) || 'Internal server error');
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    /**
     *
     * Direct Debit Create Mandate
     *
     *
     * Direct Debits are payments in M-Pesa that are initiated by the Payee alone without any Payer interaction, but permission must first be granted by the Payer. The granted permission from the Payer to Payee is commonly termed a ‘Mandate’, and M-Pesa must hold details of this Mandate.
     * The Direct Debit API set allows an organisation to get the initial consent of their customers to create the Mandate that allows the organisation to debit customer's account at an agreed frequency and amount for services rendered. After the initial consent, the debit of the account will not involve any customer interaction. The Direct Debit feature makes use of the following API calls:
     * •    Create a Direct Debit mandate
     * •    Pay a mandate
     * The customer is able to view and cancel the Direct Debit mandate from G2 menu accessible via USSD menu or the Smartphone Application.
     * @api
     * @param data
     *
     *
     *
     * @returns {Promise} Promise
     */
    Pesa.prototype.debit_create = function (data, sessionID) {
        return __awaiter(this, void 0, void 0, function () {
            var output_SessionID, res, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        output_SessionID = '';
                        if (!!sessionID) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.get_session()];
                    case 1:
                        res = _a.sent();
                        output_SessionID = res.output_SessionID;
                        _a.label = 2;
                    case 2: return [4 /*yield*/, axios_1.default({
                            method: 'post',
                            url: this.baseURL + this.TRANSACTION_ROUTES.ddc,
                            headers: {
                                Accept: 'application/json',
                                Origin: '*',
                                Authorization: 'Bearer ' + this.encrypt_key(output_SessionID),
                            },
                            data: {
                                input_AgreedTC: data.input_AgreedTC,
                                input_Country: data.input_Country,
                                input_CustomerMSISDN: data.input_CustomerMSISDN,
                                input_EndRangeOfDays: data.input_EndRangeOfDays,
                                input_ExpiryDate: data.input_ExpiryDate,
                                input_FirstPaymentDate: data.input_FirstPaymentDate,
                                input_Frequency: data.input_Frequency,
                                input_ServiceProviderCode: data.input_ServiceProviderCode,
                                input_StartRangeOfDays: data.input_StartRangeOfDays,
                                input_ThirdPartyConversationID: data.input_ThirdPartyConversationID,
                                input_ThirdPartyReference: data.input_ThirdPartyReference,
                            },
                        })];
                    case 3:
                        response = _a.sent();
                        return [2 /*return*/, __assign(__assign({}, response.data), { status: 200 })];
                }
            });
        });
    };
    /**
     * Direct Debit Payment
     *
     * Direct Debits are payments in M-Pesa that are initiated by the Payee alone without any Payer interaction, but permission must first be granted by the Payer. The granted permission from the Payer to Payee is commonly termed a ‘Mandate’, and M-Pesa must hold details of this Mandate.
     * The Direct Debit API set allows an organisation to get the initial consent of their customers to create the Mandate that allows the organisation to debit customer's account at an agreed frequency and amount for services rendered. After the initial consent, the debit of the account will not involve any customer interaction. The Direct Debit feature makes use of the following API calls:
     * •    Create a Direct Debit mandate
     * •    Pay a mandate
     * The customer is able to view and cancel the Direct Debit mandate from G2 menu accessible via USSD menu or the Smartphone Application.
     *
     * @api

     * @param data

     * @returns {Promise} Promise
     */
    Pesa.prototype.debit_payment = function (data, sessionID) {
        return __awaiter(this, void 0, void 0, function () {
            var output_SessionID, res, response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        output_SessionID = '';
                        if (!!sessionID) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.get_session()];
                    case 1:
                        res = _a.sent();
                        output_SessionID = res.output_SessionID;
                        _a.label = 2;
                    case 2: return [4 /*yield*/, axios_1.default({
                            method: 'post',
                            url: this.baseURL + this.TRANSACTION_ROUTES.ddp,
                            headers: {
                                Accept: 'application/json',
                                Origin: '*',
                                Authorization: 'Bearer ' + this.encrypt_key(output_SessionID),
                            },
                            data: {
                                input_Amount: data.input_Amount,
                                input_Country: data.input_Country,
                                input_Currency: data.input_Currency,
                                input_CustomerMSISDN: data.input_CustomerMSISDN,
                                input_ServiceProviderCode: data.input_ServiceProviderCode,
                                input_ThirdPartyConversationID: data.input_ThirdPartyConversationID,
                                input_ThirdPartyReference: data.input_ThirdPartyReference,
                            },
                        })];
                    case 3:
                        response = _a.sent();
                        return [2 /*return*/, __assign(__assign({}, response.data), { status: 200 })];
                }
            });
        });
    };
    return Pesa;
}());
exports.Pesa = Pesa;
